### PR TITLE
chore(spells): refactor spell service calls [NET-522]

### PR DESCRIPTION
**Pull request by kmd-fl**
## Description

Gather all spell service calls into one crate for easier control over used spell service interface.

## Motivation
* The code looked too ugly
* Make one entry point of control over spell service interface

## Proposed Changes
* Created new crate `spell-service-api` which provides functions over spell service for easier and unified access
* Refactor sorcerer, spell builtins and system service deployer to use it  

## Additional Notes
* This crate can help to implement support of several spell services versions
* I don't like that this crates owns the whole `PartcileAppService`, but maybe it's fine for the first iteration of refactoring. Or maybe there's an easy solution I don't see for some reason
* Any refactor suggestions are welcome

### CODE REVIEW

**Review by gurinderu — APPROVED** [review]

File: sorcerer/src/script_executor.rs [hunk 1]
```diff
@@ -128,21 +91,11 @@ impl Sorcerer {
...
-                worker_id,
-                self.spell_script_particle_ttl,
-            ),
-            &event.spell_id,
-            "set_string",
-        )
-        .map(drop)
+        let params = CallParams::local(event.spell_id, worker_id, self.spell_script_particle_ttl);
+        self.spell_service_api
+            // should it be `api.set_string` or `api.set_trigger`, so that "trigger" keyword is a part of API?
```
Comment (justprosh) [inline sorcerer/src/script_executor.rs:96]: Yeah, I think it will be nice to create `set_trigger` method

File: sorcerer/src/utils.rs [hunk 2]
```diff
@@ -20,6 +20,8 @@ use serde::de::DeserializeOwned;
 use particle_args::JError;
 use particle_execution::{FunctionOutcome, ParticleParams};
 
+// TODO: mark as such for CI to pass, remove later
```
Comment (justprosh) [inline sorcerer/src/utils.rs:23]: Don't forget to remove this func at all

File: sorcerer/src/spell_builtins.rs [hunk 3]
```diff
@@ -370,87 +301,57 @@ pub(crate) fn get_spell_id(params: ParticleParams) -> Result<JValue, JError> {
...
+            "Failed to parse argument `{key} -> {str_value}` for spell {spell_id}: {e}"
         ))
     })
 }
 
 pub(crate) fn store_error(
     mut args: Args,
     params: ParticleParams,
-    services: ParticleAppServices,
+    // services: ParticleAppServices,
```
Comment (justprosh) [inline sorcerer/src/spell_builtins.rs:325]:
```suggestion
```

File: sorcerer/src/spell_builtins.rs [hunk 4]
```diff
@@ -72,51 +74,14 @@ pub async fn install_spell(
     )?;
     spell_storage.register_spell(worker_id, spell_id.clone());
 
+    let params = CallParams::local(spell_id.clone(), worker_id, Duration::from_millis(ttl));
     // TODO: refactor these service calls
```
Comment (justprosh) [inline sorcerer/src/spell_builtins.rs:78]:
```suggestion
```

File: sorcerer/src/spell_builtins.rs [hunk 5]
```diff
@@ -13,18 +13,19 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-use fluence_spell_dtos::value::{ScriptValue, SpellValueT, StringValue, UnitValue};
+//use fluence_spell_dtos::value::{StringValue, UnitValue};
```
Comment (justprosh) [inline sorcerer/src/spell_builtins.rs:16]:
```suggestion
```

File: crates/spell-service-api/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+                    target_type: std::any::type_name::<T>(),
+                    reason: e.to_string(),
+                }),
+            },
+        }
+    }
+}
+
+/*
+#[cfg(test)]
```
Comment (justprosh) [inline crates/spell-service-api/src/lib.rs:264]: Is it supposed to be commented out?

File: crates/spell-service-api/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+        self.set_string(params, "trigger".to_string(), event)
+    }
+
+    pub fn store_error(&self, params: CallParams, args: Vec<Value>) -> Result<(), CallError> {
+        let function = Function {
+            name: "store_error",
+            args,
+        };
+        let _ = self.call::<UnitValue>(params, function);
+        Ok(())
```
Comment (justprosh) [inline crates/spell-service-api/src/lib.rs:213]: You forgot to check result value here with `?`. I suggest to write it with `drop` like this:
```
self.call::<UnitValue>(params, function).map(drop)
```

**Comment by kmd-fl** [comment]
@folex posmotri plz

**Comment by kmd-fl** [comment]
@folex :gun: posmotri or I merge it

**Review by justprosh — APPROVED** [review]

File: crates/system-services/src/deployer.rs [hunk 8]
```diff
@@ -414,28 +418,24 @@ impl Deployer {
...
+            spell_id.to_string(),
+            self.root_worker_id,
+            Duration::from_millis(DEPLOYER_TTL),
+        );
+        // update trigger config
+        self.spell_service_api
+            .set_trigger_config(params.clone(), spell_distro.trigger_config.clone())?;
+        // update spell
+        self.spell_service_api
+            .set_script(params.clone(), spell_distro.air.to_string())?;
```
Comment (folex) [inline crates/system-services/src/deployer.rs:433]: can you move trigger config and air to separate variables so it reads better?

File: crates/system-services/src/deployer.rs [hunk 9]
```diff
@@ -471,7 +471,13 @@ impl Deployer {
             // Stop old spell
             let result: eyre::Result<_> = try {
                 // Stop the spell to avoid re-subscription
-                self.call_service(&spell_id, "set_trigger_config", vec![json!(empty_config)])?;
+                let params = CallParams::local(
+                    spell_id.clone(),
+                    self.root_worker_id,
+                    Duration::from_millis(DEPLOYER_TTL),
+                );
+                self.spell_service_api
```
Comment (folex) [inline crates/system-services/src/deployer.rs:479]: maybe something like that will make it a single line?
```suggestion
                self.spells_api
```

**Pull request merged**